internal/registry/registry.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return err
+	}
+
+	return os.Rename(w.Filesystem.Root(), r.Root)
+}
+
+func (r *Cache) validateCache() error {
+	repository, err := git.PlainOpen(r.Root)
+	if err != nil {
+		return errors.Wrap(err, "could not open registry cache")
+	}
+
+	remotes, err := repository.Remotes()
+	if err != nil {
+		return errors.Wrap(err, "could not access registry cache")
+	}
+
+	if len(remotes) != 1 || len(remotes[0].Config().URLs) != 1 {
+		return errors.New("invalid registry cache remotes")
+	} else if remotes[0].Config().URLs[0] != r.URL {
+		return errors.New("invalid registry cache origin")
+	}
+	return nil
+}
+
+func (r *Cache) Initialize() error {
+	_, err := os.Stat(r.Root)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = r.createCache()
+			if err != nil {
+				return errors.Wrap(err, "could not create registry cache")
+			}
+		}
+	}
+
+	if err := r.validateCache(); err != nil {
+		err = os.RemoveAll(r.Root)
+		if err != nil {
+			return errors.Wrap(err, "could not reset registry cache")
+		}
+		err = r.createCache()
+		if err != nil {
+			return errors.Wrap(err, "could not create registry cache")
+		}
+	}
+
+	return nil
+}
+
+func (r *Cache) Refresh() error {
+	if err := r.Initialize(); err != nil {
+		return err
+	}
+
+	repository, err := git.PlainOpen(r.Root)
+	if err != nil {
+		return errors.Wrapf(err, "could not open (%s)", r.Root)
+	}
+
+	w, err := repository.Worktree()
+	if err != nil {
+		return errors.Wrapf(err, "could not read (%s)", r.Root)
+	}
+
+	err = w.Pull(&git.PullOptions{RemoteName: "origin"})
+	if err == git.NoErrAlreadyUpToDate {
+		return nil
+	}
+	return err
+}
+
+func (r *Cache) readEntry(ns, name, version string) (Entry, error) {
+	var indexDir string
+	switch {
+	case len(name) < 3:
+		indexDir = name
+	case len(name) < 4:
+		indexDir = filepath.Join(name[:2], name[2:])
+	default:
+		indexDir = filepath.Join(name[:2], name[2:4])
+	}
+
+	index := filepath.Join(r.Root, indexDir, fmt.Sprintf("%s_%s", ns, name))
+
+	if _, err := os.Stat(index); err != nil {
+		return Entry{}, errors.Wrapf(err, "could not find buildpack: %s/%s", ns, name)
+	}
+
+	file, err := os.Open(index)
+	if err != nil {
+		return Entry{}, errors.Wrapf(err, "could not open index for buildpack: %s/%s", ns, name)
+	}
+	defer file.Close()
+
+	entry := Entry{}
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var bp Buildpack
+		err = json.Unmarshal([]byte(scanner.Text()), &bp)
+		if err != nil {
+			return Entry{}, errors.Wrapf(err, "could not parse index for buildpack: %s/%s", ns, name)
+		}
+
+		entry.Buildpacks = append(entry.Buildpacks, bp)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return entry, errors.Wrapf(err, "could not read index for buildpack: %s/%s", ns, name)
+	}
+
+	return entry, nil
+}
+
+func (b *Buildpack) Validate() error {
+	if b.Address == "" {
+		return errors.New("address is a required field")
+	}
+	if _, err := ggcrname.ParseReference(b.Address, ggcrname.WeakValidation); err != nil {
```

Comment:
1. yea, but I'm not really sure how to do that in Golang. Like this will be an instance of a `Digest` in ggcr, but we only see it as a `Reference`.
2. tag and digest are mutually exclusive, so the call to `NewDigest` will error if there's a tag IIRC

Reply:
Is https://github.com/google/go-containerregistry/pull/703/files#diff-384cbc32f17b860e5335ef202256ac9eR69 relevant to this discussion?

Reply:
actually, all `ParseReference` does is call `NewDigest`, so I think i can just omit that call. https://github.com/google/go-containerregistry/blob/master/pkg/name/ref.go#L44